### ISSUE TITLE
Support package renames 

### ISSUE THREAD

**Issue by loic-sharma** [NuGet/NuGetGallery#7896]
# Background
Spec: [Support package renames](https://github.com/NuGet/Home/wiki/Support-package-renames)

# Tasks

## Minimum Viable Product
These are the highest priority work items. Once they are complete, we will have an internal MVP that our partner team can use to give feedback:

* [x] [NuGetGallery#7900](https://github.com/NuGet/NuGetGallery/issues/7900) - Add new schema
* [x] [NuGetGallery#7898](https://github.com/nuget/nugetgallery/issues/7898) - Apply popularity transfers in search
* [x] [NuGetEngineering#3087](https://github.com/NuGet/Engineering/issues/3087) - Validate MVP with internal team
* [x] [NuGetGallery#7944](https://github.com/NuGet/NuGetGallery/issues/7944) - Add feature flag to undo popularity transfers

## A/B test

Next we can A/B test the popularity transfers on PROD to ensure the new search rankings does not regress customer scenarios:

* [x] [NuGetGallery#7903](https://github.com/nuget/nugetgallery/issues/7903) - Expose popularity transfers in search
* [x] [NuGetGallery#7905](https://github.com/nuget/nugetgallery/issues/7905) - Monitor the popularity transfer freshness
* [x] [NuGetGallery#7966](https://github.com/NuGet/NuGetGallery/issues/7966) - Monitor V3 jobs' feature flag staleness
* [x] [NuGetGallery#7977](https://github.com/NuGet/NuGetGallery/issues/7977) - Unblock deployments to existing search instances
* [x] [NuGetGallery#7906](https://github.com/nuget/nugetgallery/issues/7906) - A/B test popularity transfers 

## General Release
Once we feel confident that the feature addresses the needs of our internal partner, we can work on releasing this feature to all customers. 

Add Gallery UI:
* [x] [NuGetGallery#7907](https://github.com/nuget/nugetgallery/issues/7907) - Show replacement packages on "Display Package" page
* ~~[ ] [NuGetGallery#7908](https://github.com/nuget/nugetgallery/issues/7908) - Add package renames form to "Manage Package" page~~
* ~~[ ] [NuGetGallery#7910](https://github.com/nuget/nugetgallery/issues/7910) - Add message for pending popularity transfers on "Manage Package" page~~
    * This depends on [NuGetGallery#7903](https://github.com/nuget/nugetgallery/issues/7903) - Expose popularity transfers in search
* ~~[ ] [NuGetGallery#7911](https://github.com/nuget/nugetgallery/issues/7911) - Reduce popularity transfer propagation times~~
* [ ] [NuGetGallery#7943](https://github.com/NuGet/NuGetGallery/issues/7943) - Add documentation

Testing:
* [ ] [NuGetGallery#7897](https://github.com/nuget/nugetgallery/issues/7897) - Add "categories" to NDCG data sets
* ~~[ ] [NuGetGallery#7909](https://github.com/nuget/nugetgallery/issues/7909) - Add Gallery API for end-to-end testing~~
* ~~[ ] [NuGetGallery#7904](https://github.com/nuget/nugetgallery/issues/7904) - Add end-to-end test~~
    * This depends on [NuGetGallery#7911](https://github.com/nuget/nugetgallery/issues/7911) - Reduce popularity transfer propagation times

## Future Refinements
Once the feature has been released to all customers, we can consider the following future refinements. These are low priority:
* [ ] [NuGetGallery#7901](https://github.com/nuget/nugetgallery/issues/7901) - Consider letting new packages match on old packages' ID
* [ ] [NuGetGallery#7902](https://github.com/nuget/nugetgallery/issues/7902) - Refine the transfer popularity algorithm
* [ ] [Engineering#3089](https://github.com/nuget/engineering/issues/3089) - Remove manual download overrides

# Recently Triaged Issues

All issues in this section should be triaged by the v-team into one of their business objectives or features.

- [ ] https://github.com/NuGet/NuGetGallery/issues/7897
- [ ] https://github.com/NuGet/NuGetGallery/issues/7901

**Comment by Nirmal4G** [comment]
Also introduce a concept of Package family by namespaces.

for eg.: All NuGet.* packages by **NuGet** fall under **NuGet** family but 3rd party NuGet.* packages can listed under the same home but under 3rd party packages!!

**Comment by loic-sharma** [comment]
Thanks for the suggestion @Nirmal4G! NuGet.org has package ID prefix reservation that I think may already address this concern. For example, the `NuGet.` package ID prefix is reserved to the [nuget account](https://www.nuget.org/profiles/nuget). As a result, `NuGet.*` packages owned by `nuget` account (like  [`NuGet.Packaging`](https://www.nuget.org/packages/NuGet.Packaging/5.6.0-preview.2.6489)) have a [blue checkmark](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation#prefix-reservation-visual-indicator). `NuGet.*` packages not owned by `nuget` (like [NuGet.Build.Packaging](https://www.nuget.org/packages/NuGet.Build.Packaging/)) do not have this blue checkmark. I hope that help!

**Comment by Nirmal4G** [comment]
ID prefix verifies who owns the prefix and not answering the question `Are there any alternatives or different functionalities based on NuGet by 3rd parties available?`

My solution answers to that. Basically 3rd party packages visibility. There are great solutions available by 3rd parties but they are not searchable much.

**Comment by loic-sharma** [comment]
I see. Thanks for the feedback!

**Comment by Nirmal4G** [comment]
I tried using [NuGet search syntax](https://docs.microsoft.com/en-us/nuget/consume-packages/finding-and-choosing-packages#search-syntax). I can't exclude NuGet packages by NuGet owner. Do we have `-owner:nuget` syntax for exclusion?

**Comment by loic-sharma** [comment]
We don't have any exclusion syntax today, no.

I'm curious, what kind of packages would you like to find? Would you like to find packages that "extend" functionality provided by `NuGet.*` packages? Would a "reverse dependency" feature help? See: https://github.com/NuGet/Home/wiki/Reverse-Dependencies-(Dependent-Packages)

**Comment by Nirmal4G** [comment]
Not just for NuGet but for any other package that has an extended 3rd party solution. `NuGetizer-3000` was hard find. If Miguel didn't tweet that, I wouldn't have found it, just by browsing through github/nuget. Some packages like MSBuild tools and SDKs don't have dependencies node in the metadata, since they are self contained packages. Those are hard to find. Though the new `PackageType` filter in the search help but not for regular packages.

**Comment by clairernovotny** [comment]
Does this feature address the diamond dependency issue? 

A core problem is that renaming a package will cause conflicts in a graph containing packages that reference the new name and ones that reference the old name. Assuming that the namespaces in assemblies in the package can't be changed, this will cause lots of conflicts that are unsolvable by the end user.

**Comment by loic-sharma** [comment]
After talking to internal partners we have decided to abandon the "package renames" feature. We believe it would be better to improve the package deprecation feature instead.

We will keep the popularity transfer feature and will allow customers to apply using a process similar to our [Package Id prefix reservation process](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation). /cc @chgill-MSFT

**Comment by mdell-seradex** [comment]
I hope the package renames feature is reconsidered for inclusion.
I would like to see this ability added.

**Comment by theolivenbaum** [comment]
@loic-sharma what was the rationale to abandon this feature?

**Comment by loic-sharma** [comment]
@theolivenbaum Hello, I'm no longer on the NuGet team. I don't know what the team's plans are for package renames.
